src/w3d/lib/slist.h

```diff
@@ -12,73 +13,262 @@
  *            LICENSE
  */
-#include "always.h"
-#include "autopool.h"
-template<typename T> class SList;
-
-class GenericSLNode : public AutoPoolClass<GenericSLNode, 256>
-{
-protected:
-    void *Internal_Get_Next() { return m_nodeNext; };
-    void Internal_Set_Next(void *n) { m_nodeNext = n; };
-    void *Internal_Get_Data() { return m_nodeData; };
-    void Internal_Get_Data(void *d) { m_nodeData = d; };
-
-    GenericSLNode(void *obj)
-    {
-        m_nodeData = obj;
-        m_nodeNext = nullptr;
-    };
+#pragma once
 
-private:
-    GenericSLNode(){};
-
-    void *m_nodeNext;
-    void *m_nodeData;
-};
+#include "always.h"
+#include "slnode.h"
 
-template<typename T> class SLNode : public GenericSLNode
+template<class T> class SList
 {
 public:
-    friend class SList<T>;
-
-    SLNode<T> *Next() { return reinterpret_cast<SLNode<T> *>(Internal_Get_Next()); }
-    T *Data() { return reinterpret_cast<T *>(Internal_Get_Data()); }
-    void Set_Next(SLNode<T> *n) { Internal_Set_Next(reinterpret_cast<void *>(n)); }
+    SList() : m_headNode(nullptr), m_tailNode(nullptr) {}
+    virtual ~SList() { Remove_All(); }
 
-protected:
-    SLNode(T *obj) : GenericSLNode(obj) {}
-
-private:
-    SLNode(){};
-};
+    SLNode<T> *Head() const { return m_headNode; }
+    SLNode<T> *Tail() const { return m_tailNode; }
 
-template<class T> class SList
-{
-public:
-    SList()
-    {
-        m_headNode = nullptr;
-        m_tailNode = nullptr;
-    };
-    virtual ~SList() { Remove_All(); };
-    SLNode<T> *Head() const;
-    SLNode<T> *Tail() const;
     SLNode<T> *Find_Node(T *data) const;
+
     virtual bool Add_Head(T *data);
     virtual bool Add_Head(SList<T> &list);
     virtual bool Add_Tail(T *data);
     virtual bool Add_Tail(SList<T> &list);
+
     virtual T *Remove_Head();
     virtual T *Remove_Tail();
     virtual bool Remove(T *element);
     virtual void Remove_All();
+
     virtual bool Insert_Before(T *newnode, T *oldnode = nullptr);
     virtual bool Insert_After(T *newnode, T *oldnode = nullptr);
-    virtual bool Is_Empty() const;
-    virtual long Get_Count() const;
+
+    virtual bool Is_Empty() const { return m_headNode == nullptr; }
+
+    virtual uint32_t Get_Count() const;
 
 private:
     SLNode<T> *m_headNode;
     SLNode<T> *m_tailNode;
 };
+
+template<class T> inline SLNode<T> *SList<T>::Find_Node(T *data) const
+{
+    SLNode<T> *cur;
+
+    for (cur = m_headNode; cur != nullptr && cur->Data() != data; cur = cur->Next()) {
+    }
+
+    return cur;
+}
+
+template<class T> inline bool SList<T>::Add_Head(T *data)
+{
+    if (data == nullptr) {
+        return false;
+    }
+
+    SLNode<T> *temp = new SLNode<T>(data);
+    temp->Set_Next(m_headNode);
+    m_headNode = temp;
+
+    if (m_tailNode == nullptr) {
+        m_tailNode = temp;
+    }
+
+    return true;
+}
+
+template<class T> inline bool SList<T>::Add_Head(SList<T> &list)
+{
+    if (list.m_headNode == nullptr) {
```

Comment:
I suspect all of these `if (list.m_headNode == nullptr) ` are actually `if (Is_Empty())`